WIP: Add blog article about rapid deployment of git to windows machines using PowerShell (#28)
PowerShell/Community-Blog: Posts/2021/03/DarwinJS-Contribution-LightningFastandEasyProvisioningofGitwithSSHKeyAuthenticationonWindows.md

---
post_title: Lightning Fast and Easy Provisioning of Git with SSH Key Authentication on Windows

> username: DarwinJS@sanoys.com

Reviewer: Think this needs to match your username `darwinjs`, not your email address.

Author: For all the existing articles the email is used - could you verify?

Catagories: PowerShell
tags: Git, DevOps
Summary: Getting your team setup with Git over SSH as quickly as possible!

> featured_image: ./media/DarwinJS-Contribution-LightningFastandEasyProvisioningofGitwithSSHKeyAuthenticationonWindows/windows-git-ssh.png

Reviewer: I think this will work for the featured image. We may need to scale the size but I won't be able to tell that until it gets copied to WordPress.

> CanonicalURL: https://missionimpossiblecode.io/post/lightning-fast-and-easy-provisioning-of-git-with-ssh-key-authentication-on-windows/

Reviewer: This metadata value will be ignored by the system, but I can use it to configure the canonical URL before publishing.

Author: Yes that was my intent!

---

Maybe you have a team of Windows developers that are onboarding for your new Git server installation or maybe you've decided to drop http password authentication to your existing Git server (due to it's many problems).  Your next steps may well be into a rough and rocky rabbit hole when you were holding out hope for simplicity (you know the kind you've fallen into before if you've been in tech for more than about 45 minutes).

> The guides on the internet for getting Windows setup for SSH authentication for Git are unnecessarily complex.

Reviewer: Can we simplify this sentence?

```suggestion
The guides on the internet for setting up SSH authentication for Git for Windows setup are unnecessarily complex.
```

Author: Reworded completely in another commit.

My inner tool smith really loathes when the very first steps into something new are fraught with rocky rabbit holes - so I took on the challenge of creating an easier way. 

The resultant tool is a 20 line PowerShell script that deploys Git, configures SSH and leaves the public key on your clipboard so you can paste it into GitLab or any other Git collaborative webserver. There is also an optional connectivity test.

<!-- more -->

> 

Reviewer: We don't support the summary views in our feed. So this is not used.

Author: Removed in another commit!

## Reasons For Moving to SSH

There are multiple reasons you may want to move your Windows developers to SSH authentication for Git:

1. You want to get away from git storing local passwords - whether in the git config or in Windows Credentials (with the windows credential helper) because it is pure pain to walk people through how to find and update this password when they change it on the Git server.
2. You want to avoid both http passwords and the http protocol for git.

## Conventional Wisdom on SSH Configuration

The conventional wisdom solution offers many steps that are roughly:

1. Installing git manually.
2. Installing the well known Windows SSH client Putty
3. Installing Putty's key generator.
4. Converting the non-compatible putty generated key into an ssh compatible one.
5. Precisely placing the SSH key on disk.

> 6. Precisely and ~~manually~~ permissioning the SSH key and it's parent folder (ssh is purposely fussy about this in order to keep the key secure).

Reviewer: Did you intend to use strikethrough formatting here? It was not in the original blog post. Is you want to use it then I suggest the following change.

```suggestion
6. Precisely ~~and manually~~ permissioning the SSH key and it's parent folder (ssh is purposely fussy about this in order to keep the key secure).
```

Author: This was a mistake - fixed in another commit.

> Most of this can be avoided by simply using the full SSH client that is embedded inside of the Windows git client install.

## The Cleanest Way (With Working Automation Code) 

> Besides the above pure pain, here are the additional things solved for in this code:

Reviewer: ```suggestion
Besides the pure pain described above, here are the additional things solved for in this code:
```

> 1. Automatically installs Git - but only if necessary (idempotent)

Reviewer: You use the term _idempotent_ here several times. But the general audience really doesn't understand what that means. I don't think I understand your intended meaning in this context. Is it important to include? If so, should define what it means in this context.

Author: Originally it was intended to name the phrase "but only if necessary".  I have added  a section "### Stealing Lessons From Desired State Automation" to explain more clearly and make it part of the learning of the article.

2. Automatically installs chocolatey to install Git - but only if necessary (idempotent)
3. Automatically generates an SSH key - but only if necessary (idempotent) (which avoids killing a key that might be in use)
4. Uses the Git's built-in SSH client to create SSH keys (avoids the complexity of the above conventional wisdom)
5. Copies the public key to the clip board and pauses for the user to add it to the Git server (in their profile)
6. Optionally does a SSH login test if you provide a value for: $SSHEndPointToGitForTesting

## Solution Details

This code can be run directly from GitLab with this command:

> ```bash

Reviewer: These code blocks are powershell not bash. So use the powershell label.
```suggestion
```powershell
```

Author: Yeah, at the the time this was written, the syntax highligher in the Hugo static site generator did not yet support powershell.

Invoke-Expression -command "Invoke-WebRequest -uri 'https://gitlab.com/missionimpossiblecode/MissionImpossibleCode/-/raw/master/install-gitwithssh.ps1' -UseBasicParsing -OutFile ./install-gitwithssh.ps1" ; . ./install-gitwithssh.ps1
```

If you want to download dynamically, but also want the test and instructions to work, then set these environment variables before calling the above:

> ```bash

Reviewer: ```suggestion
```powershell
```

$env:YourGitServerhttpURL="https://gitlab.com"
$env:GitSSHUserAndEndPointForTesting="git@gitlab.com" #some Git servers might want the windows userid "git", which is specified as $env:username
```

You can also simply copy the code, hardcode the two variables and distribute it in your organization.

## Main Code

> ```bash
> 
> # Set environment variables before calling in order to test

Reviewer: Remove the blank line and change the label.

```suggestion
```powershell
# Set environment variables before calling in order to test
```
